### ISSUE TITLE
automatic debug data collection without running the system out of space

### ISSUE THREAD

**Issue by davepacheco** [oxidecomputer/omicron#2478]
This ticket covers minimum target work required to (1) make sure we have basic debug data collected on all systems, while (2) not putting system availability at risk (by filling up important ZFS datasets or the pool itself).  This came out of a recorded discussion on 2023-03-03.

- [x] Make sure that on each sled we create a "debug" ZFS dataset, probably on one U.2 device, that will store crash dumps, core files, log files, and potentially other regularly-collected data.  Determine (as a matter of policy) how large we're willing to let it get and assign that as its quota.  (I assume this will be Sled Agent that does this but I'm not sure.)  (Optional: create separate child datasets for core files, dumps, logs, etc. so that we can control their quotas separately [so that a flurry of crash dumps doesn't starve log files or vice versa].)
- [x] Set up a dump device (Sled Agent?)
- [x] Configure dumpadm to save crash dumps in that dataset.  (Sled Agent?)
- [x] Configure coreadm to save core dumps in that dataset.  (Sled Agent?)
- [x] Configure cron + logadm in all control plane zones to rotate all log files in the zone into some known location.  (Probably part of the image build.)
- [x] Configure cron + logadm in the global zone to rotate all log files into the "debug" dataset. (Sled Agent? Host OS image?)
- [x] Update Sled Agent to manage the storage in the "debug" dataset according to some policy, which will presumably start very simply (e.g., delete the oldest files in the dataset until free space in the dataset reaches a watermark like 20%)

There are some known limitations of this: most notably that files are not replicated across multiple devices. Failure of the wrong U.2 (or removal from the system) means we lose the debug data from that system.  This can be mitigated in future work (e.g., by copying them to a second dataset on another pool, or copying them into some intra-rack storage system, etc.).

It's conceivable that we ship an MVP without this, but it's a pretty big risk.  Either we don't manage the storage (in which case we risk running out of disk space and disrupting service) or we turn off all these sources of data (in which case we'll have a pretty hard time fixing anything).

**Comment by jclulow** [comment]
Note that in addition to configuring the dump directory it will probably be necessary to run `savecore` with appropriate arguments, to check the dump device for an existing dump and save it out.  There is also the question of which M.2 we use for dumping (is it the BSU from which we booted?) and then how and whether to check the _other_ BSU for a prior dump we have not yet extracted.

**Comment by wesolows** [comment]
Does this cover getting the boot-time logs (including fmd state) out of the ramdisk and into persistent storage, once sled agent has created and imported that pool?  There is some dance to be done there that will involve restarting daemons etc.  Alternately, service dependencies need to be created so that services writing to /var/log and /var/fm (but *not* /var/lock or other places we really do NOT want to persist state) can't start until sled agent has done that.  There are any number of ways to solve this set of problems and I don't mean to constrain it.  This isn't strictly about support, so it could also be part of a different bug -- but this is related to how/where/when we run savecore.

**Comment by davepacheco** [comment]
> Does this cover getting the boot-time logs (including fmd state) out of the ramdisk and into persistent storage, once sled agent has created and imported that pool? There is some dance to be done there that will involve restarting daemons etc. Alternately, service dependencies need to be created so that services writing to /var/log and /var/fm (but _not_ /var/lock or other places we really do NOT want to persist state) can't start until sled agent has done that. There are any number of ways to solve this set of problems and I don't mean to constrain it. This isn't strictly about support, so it could also be part of a different bug -- but this is related to how/where/when we run savecore.

Good question.  Yes, I was assuming that Sled Agent would be responsible for periodically copying a bunch of GZ log files into the debug dataset, including various FMA error logs and things like /var/log/messages.  I was assuming (perhaps naively?) that we could rely on the existing logadm configuration that rotates those logs and then copy the rotated logs into the debug dataset.  (Note that the logadm configuration for things like the FMA error log use `fmadm rotate` to do it, which I assume takes care of any signaling of fmd that's necessary.)

**Comment by lifning** [comment]
awaiting review:
- coreadm to m.2 crash dir and then archive to debug dataset: https://github.com/oxidecomputer/omicron/pull/3677
- move logadm-rotated log files in all zones to debug dataset: https://github.com/oxidecomputer/omicron/pull/3713
- manage storage in debug dataset by deleting older archived files: https://github.com/oxidecomputer/omicron/pull/3735
